src/hypothesis/stateful.py

```diff
@@ -354,13 +353,59 @@ def precondition_wrapper(*args, **kwargs):
             setattr(precondition_wrapper, PRECONDITION_MARKER, precond)
         else:
             new_rule = Rule(targets=rule.targets, arguments=rule.arguments,
-                            function=rule.function, precondition=precond,
-                            parent_rule=rule.parent_rule)
+                            function=rule.function, precondition=precond)
             setattr(precondition_wrapper, RULE_MARKER, new_rule)
+
+        invariant = getattr(f, INVARIANT_MARKER, None)
+        if invariant is not None:
+            new_invariant = Invariant(function=invariant.function,
+                                      precondition=precond)
+            setattr(precondition_wrapper, INVARIANT_MARKER, new_invariant)
+
         return precondition_wrapper
     return decorator
 
 
+Invariant = namedtuple(
+    u'Invariant',
+    (u'function', u'precondition')
+)
+
+
+def invariant():
```

Comment:
I could be persuaded either way, but it would seem more natural for invariant to be a single level decorator given that it takes no arguments, so you use it as:

```python
@invariant
def foo(self):
    ....
```

instead of 


```python
@invariant()
def foo(self):
    ....
```

Even if we decide to add arguments to it later (and I can't think of any we would) it's not hard to do that in a backwards compatible way that preserves this usage.